### PR TITLE
Documentation for `torch.autocast`

### LINKED ISSUE #95547: Incorrect example in documentation for `autocast`
### 🚀 The feature, motivation and pitch

There is not much space for the actual design choice for users in the device type of `autocast`. The rule is simple: use the device type of the underlying tensor operations.

~~Another consideration is the backward compatibility. The older API (1.7.0, for example) for automp autocast only accepts a `enabled` parameter, while there is a **required** parameter `device_type` in the new API, making them incompatible with each other. Users have to write an ugly if-else to support multiple pytorch APIs.~~

### Alternatives

1. At least, the examples on the page [`https://pytorch.org/docs/stable/amp.html`](https://pytorch.org/docs/stable/amp.html) need to be updated with `device_type`. The documentation is inconsistent by the time of writing.
~~2. If the proposal above is not possible due to technical reasons, a default value of `"cuda"` to `device_type` can be added to enhance compatibility between PyTorch versions.~~

Update: `torch.cuda.amp.autocast` is the old API and `torch.autocast` is the new API. I have been confusing them with each other. However, I would like to say that they *are* confusing.

### Additional context

_No response_

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by eliphatfs**
- [x] Corrected examples for CUDA devices.
- [x] Information about availability of `torch.autocast`.

Fixes #95547

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: torch/amp/autocast_mode.py [hunk 1]
```diff
@@ -170,7 +170,8 @@ def forward(self, x):
     (see :ref:`Working with Multiple GPUs<amp-multigpu>`).
 
     Args:
-        device_type(str, required):  Whether to use 'cuda' or 'cpu' device
+        device_type(str, required):  Whether to use 'cuda' or 'cpu' device.
+                                     The type can be obtained by `device.type`.
```
Comment (ptrblck) [inline torch/amp/autocast_mode.py:174]: I think `device.type` would need more explanation as it's unclear to me where this object should come from.

Comment (eliphatfs) [reply torch/amp/autocast_mode.py:174]: It's like `some_tensor.device.type`.

File: torch/amp/autocast_mode.py [hunk 2]
```diff
@@ -170,7 +170,9 @@ def forward(self, x):
     (see :ref:`Working with Multiple GPUs<amp-multigpu>`).
 
     Args:
-        device_type(str, required):  Whether to use 'cuda' or 'cpu' device
+        device_type(str, required):  Whether to use 'cuda' or 'cpu' device.
```
Comment (leslie-fang-intel) [inline torch/amp/autocast_mode.py:173]: Actually, there are more than `cuda` and `cpu` are supported for now, refer to this: https://github.com/pytorch/pytorch/blob/95621b3c2e8bc0f1d7f88fb0bcb3e3e56a0ff961/torch/amp/autocast_mode.py#L193-L218.

Comment (eliphatfs) [reply torch/amp/autocast_mode.py:173]: I noticed that the error message on L218 in the snippet you posted is also out-dated.

**Comment by eliphatfs** [comment]
CI is failing for some reason,
```plain
RuntimeError: Command `git -C /work merge-base origin/master 030a6d3fe98e46c82cdbae9b93a72ceab4febfd6` returned non-zero exit code 128

fatal: Not a valid object name origin/master
```
and
```plain
bash: ./.ci/pytorch/python_doc_push_script.sh: No such file or directory
Error: Process completed with exit code 127.
```
It doesn't look like it is my problem.

**Comment by eliphatfs** [comment]
Hi, can anyone have a look at this PR? Thanks in advance.

**Review by leslie-fang-intel — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @eliphatfs, sorry for the later response. LGTM except some conflicts you need to fix.
Also may still need @ptrblck to take a look, since most of the changes are Autocast with CUDA device.

**Comment by eliphatfs** [comment]
Hi and I have fixed the conflicts.

**Comment by kit1980** [comment]
@eliphatfs mind resolving conflicts one more time?

**Comment by eliphatfs** [comment]
I think that should be resolved now.

**Review by kit1980 — APPROVED** [review]

**Pull request closed without merging**